book/src/dev/rfcs/0012-value-pools.md

```diff
@@ -0,0 +1,497 @@
+
+#### Serialization of `ValueBalance<C>`
+
+In order to save `ValueBalance` into the disk database we must implement `IntoDisk` and `FromDisk` for `ValueBalance` and for `Amount`:
+
+```rust
+impl IntoDisk for ValueBalance<C> {
+    type Bytes = [u8; 32];
+
+    fn as_bytes(&self) -> Self::Bytes {
+        [self.transparent.to_bytes(), self.sprout.to_bytes(),
+        self.sapling.to_bytes(), self.orchard.to_bytes()].concat()
+    }
+}
+
+impl FromDisk for ValueBalance<C> {
+    fn from_bytes(bytes: impl AsRef<[u8]>) -> Self {
+        let array = bytes.as_ref().try_into().unwrap();
+        ValueBalance {
+            transparent: i64::from_be_bytes(array[0..8]).try_into().unwrap()
+            sprout: i64::from_be_bytes(array[8..16]).try_into().unwrap()
+            sapling: i64::from_be_bytes(array[16..24]).try_into().unwrap()
+            orchard: i64::from_be_bytes(array[24..32]).try_into().unwrap()
+        }
+    }
+}
+
+impl IntoDisk for Amount {
+    type Bytes = [u8; 8];
+
+    fn as_bytes(&self) -> Self::Bytes {
+        self.to_bytes()
+    }
+}
+
+impl FromDisk for Amount {
+    fn from_bytes(bytes: impl AsRef<[u8]>) -> Self {
+        let array = bytes.as_ref().try_into().unwrap();
+        i64::from_be_bytes(array).try_into().unwrap()
+    }
+}
+```
+
+#### Changes to `zebra-state/src/request.rs`
+
+Add a new field to `FinalizedState`:
+
+```rust
+pub struct FinalizedBlock {
+    ..
+    /// The value balance for transparent, sprout, sapling and orchard
+    /// inside all the transactions of this block.
+    pub(crate) block_value_balance: ValueBalance<NegativeAllowed>,
+}
+```
+
+Populate it when `PreparedBlock` is converted into `FinalizedBlock`:
+
+```rust
+impl From<PreparedBlock> for FinalizedBlock {
+    fn from(prepared: PreparedBlock) -> Self {
+        let PreparedBlock {
+            ..
+            block_value_balance,
+        } = prepared;
+        Self {
+            ..
+            block_value_balance,
+        }
+    }
+}
+```
+
+#### Changes to `zebra-state/src/service/finalized_state.rs`
+
+First we add a column for all the pools: transparent, sprout, sapling, orchard:
+
+```rust
+rocksdb::ColumnFamilyDescriptor::new("value_pools", db_options.clone()),
+```
+
+At block commit(`commit_finalized_direct()`) we create the handles for the new columns:
+
+```rust
+let value_pools = self.db.cf_handle("value_pools").unwrap();
+```
+
+Next we save each value pool into the field for each upcoming block except for the genesis block:
+
+```rust
+// Consensus rule: The block height of the genesis block is 0
+// https://zips.z.cash/protocol/protocol.pdf#blockchain
+if height == block::Height(0) {
+    batch.zs_insert(value_pools, height, ValueBalance::default());
+} else {
+    let current_pool = self.get_pool();
+    batch.zs_insert(value_pools, height, (current_pool + finalized.block_value_balance)?);
+}
+```
+
+The `get_pool()` function will get the value of the pool at the tip as follows:
```

Comment:
Rust style: we avoid "get" and just use the name of the thing we're getting.

https://github.com/rust-lang/api-guidelines/blob/9c11f67df3dd0a5d8b4720e4358c68b9dee5c9b0/src/naming.md#getter-names-follow-rust-convention-c-getter

Reply:
https://github.com/ZcashFoundation/zebra/pull/2430/commits/5ae19fc0d76e060313f03a99fe97a07740451522